modules/phase_field/src/userobjects/PolycrystalCircles.C

```diff
@@ -0,0 +1,189 @@
+/****************************************************************/
+/* MOOSE - Multiphysics Object Oriented Simulation Environment  */
+/*                                                              */
+/*          All contents are licensed under LGPL V2.1           */
+/*             See LICENSE for full restrictions                */
+/****************************************************************/
+
+#include "PolycrystalCircles.h"
+//#include "IndirectSort.h"
+#include "MooseMesh.h"
+#include "MooseVariable.h"
+#include "NonlinearSystemBase.h"
+
+template <>
+InputParameters
+validParams<PolycrystalCircles>()
+{
+  InputParameters params = validParams<PolycrystalUserObjectBase>();
+  params.addClassDescription(
+      "Polycrystal circles generated from a vector input or read from a file");
+  params.addParam<bool>(
+      "read_from_file", false, "Set to true to read the position and radius vectors from a file rather than inputing them manually");
+  params.addParam<bool>(
+      "columnar_3D", false, "3D microstructure will be columnar in the z-direction?");
+  params.addParam<std::vector<Real> >("x_positions", "x coordinate for each circle center");
+  params.addParam<std::vector<Real> >("y_positions", "y coordinate for each circle center");
+  params.addParam<std::vector<Real> >("z_positions", "z coordinate for each circle center");
+  params.addParam<std::vector<Real> >("radii", "The radius for each circle");
+  params.addParam<FileName>("file_name", "File containing circle centers and radii");
+  params.addParam<unsigned int>("header_length", 0, "Number of lines to skip at top of file");
+
+  return params;
+}
+
+PolycrystalCircles::PolycrystalCircles(const InputParameters & parameters)
+  : PolycrystalUserObjectBase(parameters),
+    _columnar_3D(getParam<bool>("columnar_3D")),
+    _grain_num(0)
+{
+}
+
+void
+PolycrystalCircles::getGrainsBasedOnPoint(const Point & point,
+                                          std::vector<unsigned int> & grains) const
+{
+  unsigned int n_grains = _centerpoints.size();
+  unsigned int cnt = 0;
+  std::vector<bool> is_at_point;
+  is_at_point.resize(n_grains);
+
+  for (unsigned int i = 0; i < n_grains; ++i)
+  {
+    Real d_x = (point(0) - _centerpoints[i](0)) * (point(0) - _centerpoints[i](0));
+    Real d_y = (point(1) - _centerpoints[i](1)) * (point(1) - _centerpoints[i](1));
+    Real d_z = (point(2) - _centerpoints[i](2)) * (point(2) - _centerpoints[i](2));
+    Real distance = 0;
+
+    if (_columnar_3D)
+      distance = std::sqrt(d_x + d_y);
+    else
+      distance = std::sqrt(d_x + d_y + d_z);
+
+    if (distance < _radii[i])
+    {
+      is_at_point[i] = true;
+      ++cnt;
+    }
+    else
+      is_at_point[i] = false;
```

Comment:
Do you want more than one grain to be at a single point? If not, you don't need the complication of building this "is_at_point" vector and then checking it again.